pkg/parse/run.go

```diff
@@ -280,63 +301,98 @@ func read(ctx context.Context, p Parser, trigger string, state *reconcilerState,
+	absHydratedRoot, err := cmpath.AbsoluteOS(options.HydratedRoot)
 	if err != nil {
 		hydrationStatus.message = RenderingFailed
 		hydrationStatus.errs = status.InternalHydrationError(err, "hydrated-dir must be an absolute path")
-		return hydrationStatus, sourceStatus
+		return srcState, hydrationStatus
 	}
 
 	var hydrationErr hydrate.HydrationError
 	if _, err := os.Stat(absHydratedRoot.OSPath()); err == nil {
-		sourceState, hydrationErr = opts.readHydratedDir(absHydratedRoot, opts.reconcilerName, sourceState)
+		srcState, hydrationErr = options.readHydratedDir(absHydratedRoot, options.reconcilerName, srcState)
 		if hydrationErr != nil {
 			hydrationStatus.message = RenderingFailed
 			hydrationStatus.errs = status.HydrationError(hydrationErr.Code(), hydrationErr)
-			return hydrationStatus, sourceStatus
+			return srcState, hydrationStatus
 		}
 		hydrationStatus.message = RenderingSucceeded
 	} else if !os.IsNotExist(err) {
 		hydrationStatus.message = RenderingFailed
 		hydrationStatus.errs = status.InternalHydrationError(err, "unable to evaluate the hydrated path %s", absHydratedRoot.OSPath())
-		return hydrationStatus, sourceStatus
+		return srcState, hydrationStatus
+	} else {
+		// Source of truth does not require hydration, but hydration-controller is running
+		hydrationStatus.message = RenderingNotRequired
+		hydrationStatus.requiresRendering = false
+		err := hydrate.NewTransientError(fmt.Errorf("rendering enabled but not required"))
+		hydrationStatus.errs = status.HydrationError(err.Code(), err)
+		return srcState, hydrationStatus
+	}
+	return srcState, hydrationStatus
+}
+
+// readFromSource reads the source or hydrated configs, checks whether the sourceState in
+// the cache is up-to-date. If the cache is not up-to-date, reads all the source or hydrated files.
+// readFromSource returns the rendering status and source status.
+func readFromSource(ctx context.Context, p Parser, trigger string, recState *reconcilerState, srcState sourceState) (renderingStatus, sourceStatus) {
+	options := p.options()
+	start := time.Now()
+
+	hydrationStatus := renderingStatus{
+		commit:            srcState.commit,
+		requiresRendering: options.renderingEnabled,
+	}
+	srcStatus := sourceStatus{
+		commit: srcState.commit,
+	}
+
+	if options.renderingEnabled {
+		srcState, hydrationStatus = readHydrationStatus(p, srcState, hydrationStatus)
+		if hydrationStatus.errs != nil {
+			return hydrationStatus, srcStatus
+		}
 	} else {
 		hydrationStatus.message = RenderingSkipped
 	}
 
-	if sourceState.syncDir == state.cache.source.syncDir {
-		return hydrationStatus, sourceStatus
+	if srcState.syncDir == recState.cache.source.syncDir {
+		return hydrationStatus, srcStatus
 	}
 
-	klog.Infof("New source changes (%s) detected, reset the cache", sourceState.syncDir.OSPath())
+	// Read all the files under state.syncDir
+	srcStatus.errs = options.readConfigFiles(&srcState)
+
+	if !options.renderingEnabled {
```

Comment:
Can this be combined with the `if options.renderingEnabled ... else` above?

Reply:
No it cannot, because this relies on `srcState.files` being populated by `readConfigFiles`. In a previous iteration I called `readConfigFiles` twice (once before `readHydrationStatus` and once after) so that I could consolidate this logic into a single if/else, but it was a bit less efficient because the filesystem tree was being walked multiple times

Reply:
Side note -`readConfigFiles()` comments might be outdated. I didn't see `rendered` in the function. 
```go
// readConfigFiles reads all the files under state.syncDir and sets state.files.
// - if rendered is true, state.syncDir contains the hydrated files.
// - if rendered is false, state.syncDir contains the source files.
```

Reply:
The description is still somewhat accurate - there isn't a `rendered` boolean but the files that it ingests are either hydrated or source files depending on the list of file paths that are passed in (I'm not aware of this changing recently). I can update the description to make this more clear